tests/test_other.py

```diff
@@ -9068,9 +9068,9 @@ def test_minimal_runtime_code_size(self):
     test_cases = [
       (asmjs + opts, hello_world_sources, {'a.html': 985, 'a.js': 289, 'a.asm.js': 113, 'a.mem': 6}),
       (opts, hello_world_sources, {'a.html': 972, 'a.js': 624, 'a.wasm': 86}),
-      (asmjs + opts, hello_webgl_sources, {'a.html': 885, 'a.js': 4980, 'a.asm.js': 10965, 'a.mem': 321}),
-      (opts, hello_webgl_sources, {'a.html': 861, 'a.js': 5046, 'a.wasm': 8842}),
-      (opts, hello_webgl2_sources, {'a.html': 861, 'a.js': 6182, 'a.wasm': 8842}) # Compare how WebGL2 sizes stack up with WebGL 1
+      (asmjs + opts, hello_webgl_sources, {'a.html': 885, 'a.js': 4974, 'a.asm.js': 10965, 'a.mem': 321}),
+      (opts, hello_webgl_sources, {'a.html': 861, 'a.js': 5040, 'a.wasm': 8842}),
+      (opts, hello_webgl2_sources, {'a.html': 861, 'a.js': 6286, 'a.wasm': 8842}) # Compare how WebGL2 sizes stack up with WebGL 1
```

Comment:
Great to see that WebGL 1 code size drops by `6` bytes, but WebGL 2 code size increases by `104` bytes, which is a `+1.02%` size creep for WebGL 2 users who are not using WebGL 2.0 compute. It would be great to not increase code size for WebGL 2 users.

Looks like the size increase is related to context creation, with `majorVersion` and `minorVersion` fields. I wonder if we should replace those with a single `featureLevel` field that would be `10`, `20` or `30` depending on the version?

Or could we get away with one version number `1`,`2`,`3` at runtime? That would be even simpler?

Reply:
> Looks like the size increase is related to context creation, with majorVersion and minorVersion fields. I wonder if we should replace those with a single featureLevel field that would be 10, 20 or 30 depending on the version?

It sounds like the feature level calculation should be done before entering [`GL.createContext()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_webgl.js#L464) to avoid passing `minorVersion` around? It seems I have to duplicate the logic in [`Browser.createContext()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_browser.js#L303) and [`emscripten_webgl_create_context()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_html5.js#L2221) then. Or am I misunderstanding anything?

> Or could we get away with one version number 1,2,3 at runtime? That would be even simpler?

Do you mean substituting `10`/`20`/`30` with `1`/`2`/`3` (sorry I didn't quite get it)? But here's my original intention:

* I think feature level 1/2/3 and WebGL version 1/2 are more likely to be misused interchangeably, or cause further confusion because they look quite similar.
* It leaves the opportunity open for inserting new feature levels between existing ones, say feature level `21` for WebGL 2.1, or feature level `19` for "almost WebGL 2.0, but [some entry points are not supported](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_webgl.js#L786)".

Reply:
>     * or feature level `19` for "almost WebGL 2.0, but [some entry points are not supported](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_webgl.js#L786)".

This is actually a good example of something we will not want to be a feature level. I.e. if Chrome "WebGL 2.0 without garbage free entry points" was featureLevel 19, then it would not be possible to easily opt out from supporting those, if one wants to target WebGL 10-20, but leave out the support for that one bad Chrome version (which nobody really is using any more). I am looking to conditionalize that Chrome version check out behind an #ifdef soon, to be handled similar to `#if WORKAROUND_OLD_WEBGL_UNIFORM_UPLOAD_IGNORED_OFFSET_BUG` is.

Reply:
> It sounds like the feature level calculation should be done before entering [`GL.createContext()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_webgl.js#L464) to avoid passing `minorVersion` around?

That would be good: having one `featureLevel` identifier that gets carried around.

>     * It leaves the opportunity open for inserting new feature levels between existing ones, say feature level `21` for WebGL 2.1

What feature level is WebGL 2.0 Compute? Is it `21` or `30`? Is there intent that a WebGL 2.1 would come about that is different from WebGL 2.0 Compute in the future? Or is there intent that a WebGL 3.0 would come about that is different from WebGL 2.0 Compute in the future?

Reply:
Due to the limitations of Apple OpenGL drivers (GL 4.1+extensions) and the existence of WebGPU, it's not very likely that we would see a WebGL 2.1 or 3.0. But I would think that 2.1 would be more likely (e.g. if we bring some common extensions into core).

Reply:
@juj I have no strong motivation to replace `supportsWebGL2EntryPoints` with feature level at the moment, so don't panic. But I haven't totally understood why feature level 19 cannot be easily opt out. Is there any obstacle to put the code path for feature level 19 in a `#if` block? Maybe I need a few lines of sample code to better understand your proposal.

WebGL 2.0 Compute is mapped to feature level 30, and the mapping is documented in `settings.js`. Mostly I agree with @kainino0x, but I guess no one would like to give an authoritative announcement for the future fate of WebGL, so this causes some uncertainty when assigning a proper feature level for WebGL 2.0 Compute.

> That would be good: having one featureLevel identifier that gets carried around.

But it looks not so good to me if the mapping logic between WebGL version and feature level will be duplicated in [`Browser.createContext()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_browser.js#L303) and [`emscripten_webgl_create_context()`](https://github.com/emscripten-core/emscripten/blob/e9c44b413781c0cdf3293bcc50894d9984930b5b/src/library_html5.js#L2221).

Reply:
Can we just stick with a simple numbering scheme here (either 10/20/21 or 1/2/3 is good with me)? I don't see any reason this can't just be changed in the future if there turns out to be a WebGL 2.1 or 3.0.